usr/lib/common/key.c

```diff
@@ -1660,6 +1663,133 @@ CK_RV rsa_priv_unwrap_get_data(TEMPLATE *tmpl,
     return CKR_OK;
 }
 
+CK_RV ibm_dilithium_priv_wrap_get_data(TEMPLATE *tmpl,
+                                       CK_BBOOL length_only,
+                                       CK_BYTE **data, CK_ULONG *data_len)
+{
+    CK_ATTRIBUTE *keyform, *rho = NULL, *seed = NULL;
+    CK_ATTRIBUTE *tr = NULL, *s1 = NULL, *s2 = NULL;
+    CK_ATTRIBUTE *t0 = NULL, *t1 = NULL;
+    CK_ATTRIBUTE *opaque = NULL;
+    CK_RV rc;
+
+    /* A private Dilithium key must have a keyform value */
+    if (!template_attribute_find(tmpl, CKA_IBM_DILITHIUM_KEYFORM, &keyform)) {
+        TRACE_ERROR("Could not find CKA_IBM_DILITHIUM_KEYFORM for the key.\n");
+        return CKR_TEMPLATE_INCOMPLETE;
+    }
+
+    /* Check if it's an expected keyform */
+    if (*(CK_ULONG *) keyform->pValue != IBM_DILITHIUM_KEYFORM_ROUND2) {
+        TRACE_ERROR("This key has an unexpected CKA_IBM_DILITHIUM_KEYFORM: %ld \n",
+                    *(CK_ULONG *) keyform->pValue);
+        return CKR_TEMPLATE_INCONSISTENT;
+    }
+
+    /* Check if key given as opaque blob */
+    if (template_attribute_find(tmpl, CKA_IBM_OPAQUE, &opaque) == FALSE) {
+
+        if (template_attribute_find(tmpl, CKA_IBM_DILITHIUM_RHO, &rho) == FALSE) {
+            TRACE_ERROR("Could not find CKA_IBM_DILITHIUM_RHO for the key.\n");
+            return CKR_FUNCTION_FAILED;
```

Comment:
Shouldn't this be CKR_TEMPLATE_INCONSISTENT (also for the following attributes)?

Reply:
yes.